### PR TITLE
Criu mips support

### PR DESCRIPTION
**Pull request by sunny868**
support arch mips base linux4.9.

### CODE REVIEW

File: compel/arch/mips/src/lib/infect.c [hunk 1]
```diff
@@ -0,0 +1,218 @@
...
+#include "common/err.h"
+#include "asm/infect-types.h"
+#include "ptrace.h"
+#include "infect.h"
+#include "infect-priv.h"
+#include "log.h"
+#include <stdio.h>
+/*
+ * Injected syscall instruction
+ * mips64el 是64位小端字节序
```
Comment (mihalicyn) [inline compel/arch/mips/src/lib/infect.c]: Please, translate to English

**Comment by sunny868** [comment]
@mihalicyn 
I do not know what I need to do , can you give me some suggest? I hope my code can be merged. thank you.

File: compel/arch/mips/src/lib/include/uapi/asm/infect-types.h [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+	/* Saved fpu registers. */
+	__u64 regs[32];
+
+	__u32 fpu_fcr31;
+	__u32 fpu_id;
+
+
+} user_fpregs_struct_t;
+
+#define MIPS_a0			regs[4] //参数寄存器 a0-a3
```
Comment (mihalicyn) [inline compel/arch/mips/src/lib/include/uapi/asm/infect-types.h]: english please

File: compel/arch/mips/src/lib/include/uapi/asm/infect-types.h [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+#define MIPS_t0			regs[8] //临时寄存器 t0-t7
```
Comment (mihalicyn) [inline compel/arch/mips/src/lib/include/uapi/asm/infect-types.h]: same

File: compel/arch/mips/src/lib/include/uapi/asm/infect-types.h [hunk 4]
```diff
@@ -0,0 +1,66 @@
...
+
+#define NATIVE_MAGIC	0x0A
+#define COMPAT_MAGIC	0x0C
+static inline bool user_regs_native(user_regs_struct_t *pregs)
+{
+	return true;
+}
+
+
+#define REG_RES(regs)		((regs).MIPS_v0) //系统调用的返回值
```
Comment (mihalicyn) [inline compel/arch/mips/src/lib/include/uapi/asm/infect-types.h]: same

File: compel/arch/mips/src/lib/include/uapi/asm/infect-types.h [hunk 5]
```diff
@@ -0,0 +1,66 @@
...
+#define REG_IP(regs)		((regs).cp0_epc) // 程序计数器
```
Comment (mihalicyn) [inline compel/arch/mips/src/lib/include/uapi/asm/infect-types.h]: same

File: compel/arch/mips/src/lib/include/uapi/asm/infect-types.h [hunk 6]
```diff
@@ -0,0 +1,66 @@
...
+#define REG_SP(regs)		((regs).MIPS_sp)
+#define REG_SYSCALL_NR(regs)	((regs).MIPS_v0) //系统调用号
```
Comment (mihalicyn) [inline compel/arch/mips/src/lib/include/uapi/asm/infect-types.h]: .

**Comment by mihalicyn** [comment]
> @mihalicyn
> I do not know what I need to do , can you give me some suggest? I hope my code can be merged. thank you.

Hi @sunny868!
The scheme is pretty simple - you post your code, then wait-wait-wait for review. Then you probably receive some comments about your code from other developers, then again wait-wait-wait. ;)

Technically, your code now looks good, and checks passed. But it's minimal condition for merging your code. It doesn't sufficient conditions. Sufficient condition is to ack several positive reviews from developers. Your changes are not simple - and it may take a long time before your changes will be accepted and merged.

**Comment by avagin** [comment]
@sunny868 Great job! Now you need to proper format these changes. The major thing what you need to do is to split these changes into a series of patches.  The idea of this is to separate each logical change into a separate patch. It is much easier to read and review small patches. Here is an example how this was done for s390 https://patchwork.criu.org/series/1716/.

Here is another useful link which describe how to split and describe changes:
 https://www.kernel.org/doc/html/v4.17/process/submitting-patches.html#describe-your-changes

Just in case if you never use git commit --interactive and git rebase --interactive, you will find them very useful to split changes into patches:
https://git-scm.com/book/en/v2/Git-Tools-Interactive-Staging

Let me know if you will have any question.

**Comment by sunny868** [comment]
@avagin I register an account in  https://patchwork.criu.org . but my email 450683921@qq.com can not get any reply info.

**Comment by avagin** [comment]
@sunny868 you don't need an account there.

**Comment by mihalicyn** [comment]
@sunny868 Maybe I'm wrong but code in `criu/arch/mips/bitops.c` file is kernel code that contains a lot of garbage code (from usermode point of view) and, again, I'm not sure, but looks like that you even not redefined `raw_local_irq_save` functions and so on. And this code should not work in userspace. Looks like these files not really used anywhere in your code and should not be contained in patchset.

Please, feel free to correct me if I'm wrong ;)

**Comment by sunny868** [comment]
@mihalicyn criu/arch/mips/bitops.c  include/common/arch/mips/asm/bitopts_mips.h include/common/arch/mips/asm/irqflags*.h are really useful now.but they are really not must in user space. raw_local_irq_save is unusefully . I will have a good change and test. thanks for you suggest.

**Comment by mihalicyn** [comment]
@sunny868 you are welcome. Please, feel free to ask questions!

File: include/common/arch/mips/asm/atomic.h [hunk 7]
```diff
@@ -0,0 +1,995 @@
+#ifndef __CR_ATOMIC_H__
```
Comment (mihalicyn) [inline include/common/arch/mips/asm/atomic.h]: Is this file needed?

File: include/common/arch/mips/asm/cmpxchg.h [hunk 8]
```diff
@@ -0,0 +1,357 @@
+#ifndef __CR_CMPXCHG_H__
```
Comment (mihalicyn) [inline include/common/arch/mips/asm/cmpxchg.h]: Is this file needed?

File: criu/arch/mips/crtools.c [hunk 9]
```diff
@@ -0,0 +1,347 @@
...
+	UserMipsRegsEntry *gpregs;
+	UserMipsFpregsEntry *fpregs;
+
+	ti_mips = xmalloc(sizeof(*ti_mips));
+	if (!ti_mips)
+		goto err;
+	thread_info_mips__init(ti_mips);
+	core->ti_mips = ti_mips;
+
+	gpregs = xmalloc(sizeof(*gpregs));
```
Comment (mihalicyn) [inline criu/arch/mips/crtools.c]: if (!gpregs) ?

File: criu/arch/mips/crtools.c [hunk 10]
```diff
@@ -0,0 +1,347 @@
...
+	user_mips_regs_entry__init(gpregs);
+	ti_mips->gpregs = gpregs;
+
+	fpregs = xmalloc(sizeof(*fpregs));
+	if (!fpregs)
+		goto err;
```
Comment (mihalicyn) [inline criu/arch/mips/crtools.c]: xfree(ti_mips) ?

File: compel/src/lib/infect.c [hunk 11]
```diff
@@ -837,6 +837,36 @@ void compel_relocs_apply(void *mem, void *vbase, size_t size, compel_reloc_t *el
...
+#ifdef CONFIG_MIPS
+		} else if (elf_relocs[i].type & COMPEL_TYPE_MIPS_26) {
+		    int *where = (mem + elf_relocs[i].offset);
+		    *where = *where | ((elf_relocs[i].addend + ((unsigned long)vbase & 0x00fffffff) /*low 28 bit*/)>>2);
+		} else if (elf_relocs[i].type & COMPEL_TYPE_MIPS_HI16) {
+		      //refer to binutils mips.cc
+		    int *where = (mem + elf_relocs[i].offset);
+
+		    int v_lo16 = (unsigned long)vbase &0x00ffff;
+		    if((v_lo16+elf_relocs[i].value+elf_relocs[i].addend) >= 0x8000){ //有符号扩展
```
Comment (mihalicyn) [inline compel/src/lib/infect.c]: comments

**Comment by sunny868** [comment]
Why CI failed?

**Comment by adrianreber** [comment]
> Why CI failed?

CI looks good. The only failure is the podman test, because the package we install is broken. This is known and should be fixed really soon.

**Comment by avagin** [comment]
The comment https://github.com/checkpoint-restore/criu/pull/871#issuecomment-561283577 is still valid. Let me know if you have any questions.

**Comment by sunny868** [comment]
@avagin thank you . I hope my PR can be merged to branch crib-dev very much. I will goon to support mips platform if there have any problem or bug. now I am testing .

**Comment by cyrillos** [comment]
@sunny868 Andrew @avagin is still requesting you to reorder patches and make the series more suitable for merging. Look at how it is done for s390.

https://patchwork.criu.org/series/1716/

First they've modified compel engine and then introduced new arch in a small steps. You could refer to this series I think and do the same.

**Comment by sunny868** [comment]
@cyrillos @avagin sorry I understood it wrong before. but I am not understand it clearly. I can split split these changes into a series of patches just like s390, then do I submit some *.patch file or PRs?

**Comment by cyrillos** [comment]
Yes, split the changes into patches and make PR

**Comment by sunny868** [comment]
@mihalicyn @cyrillos I have create a new branch loongson-criu-dev-v1 and split the changes into patches .like this:

commit 5fb3ea9a163d1deddf9aa2681dcd4f9b46b849a2 (HEAD -> loongson-criu-dev-v1, origin/loongson-criu-dev-v1)
Author: GuoYun Sun <sunguoyun@loongson.cn>
Date:   Sun Dec 22 10:58:36 2019

    mips: Enable mips build in Makefile
    
    Signed-off-by: GuoYun Sun <sunguoyun@loongson.cn>

commit 3f666be3aa85228b4a188e1e9bef60f8d80152e7
Author: GuoYun Sun <sunguoyun@loongson.cn>
Date:   Sun Dec 22 10:53:37 2019

    mips:criu: Enable mips in criu
    
    Signed-off-by: GuoYun Sun <sunguoyun@loongson.cn>

commit fe21ba9299971577a0056f33fc771f8528cfa069
Author: GuoYun Sun <sunguoyun@loongson.cn>
Date:   Sun Dec 22 10:39:30 2019

    mips:criu/arch/mips: Add mips parts to criu
    
    Signed-off-by: GuoYun Sun <sunguoyun@loongson.cn>

commit 210c69c1601e818cba842cf20d6ff388bd557169
Author: GuoYun Sun <sunguoyun@loongson.cn>
Date:   Sun Dec 22 10:27:39 2019

    mips:proto: Add mips to protocol buffer files
    
    Signed-off-by: GuoYun Sun <sunguoyun@loongson.cn>

commit dcc90b86f0be10ff214b4bf9abef320b521c69cc
Author: Guoyun Sun <sunguoyun@loongson.cn>
Date:   Sat Dec 21 16:26:49 2019

    mips:compel: Enable mips in compel/
    
    Change-Id: Iaa33194fc874a8da16822ea3c052a1d4827c4499
    Signed-off-by: Guoyun Sun <sunguoyun@loongson.cn>

commit b7b5e20b0b09ae5e9df194729ad5ec5ada079020
Author: Guoyun Sun <sunguoyun@loongson.cn>
Date:   Sat Dec 21 16:14:54 2019

    mips:compel/arch/mips: Add architecture support to compel tool and libraries
    
    This patch only adds the support but does not enable it for building.
    
    Change-Id: Ied41a5f5079f24cedb7c7c0b1826be9e715644fc
    Signed-off-by: Guoyun Sun <sunguoyun@loongson.cn>

fork criu addr: https://github.com/sunny868/criu.git 
test criu addr: https://www.travis-ci.org/sunny868/criu. 



Then ,do I have a new PR base loongson-criu-dev-v1?

**Comment by cyrillos** [comment]
yes

**Comment by sunny868** [comment]
@cyrillos @adrianreber I have a new pull request #894 ,but CI failed for TR_ARCH=openj9-test. I don't know what I can work for it

**Comment by adrianreber** [comment]
> @cyrillos @adrianreber I have a new pull request #894 ,but CI failed for TR_ARCH=openj9-test. I don't know what I can work for it

Don't worry about the openj9-test. We are currently trying to get the test working again. It is failing everywhere and not related to your change. I believe it is related to https://bugs.launchpad.net/ubuntu/+source/linux/+bug/1857257.
